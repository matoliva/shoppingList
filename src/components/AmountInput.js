import { useState, useEffect } from "react";
import { BsFillCaretLeftFill, BsFillCaretRightFill } from "react-icons/bs";
import PropTypes from 'prop-types';

import "./amountInput.css";

const AmountInput = ({quantity = 0, onSelectQuantity}) => {
    const [_quantity, setQuantity] = useState(quantity);

    const handleClickPlus = () => {
        setQuantity(_quantity + 1);
    }

    const handleClickLess = () => {
        if (_quantity > 0) {
            setQuantity(_quantity - 1);
        }
    }

    useEffect(() => {
        onSelectQuantity(_quantity);
    }, [_quantity]);

    return (
        <div className="amount-input">
            <button id="btn-subtract" className="amount-input__button" onClick={handleClickLess}> <BsFillCaretLeftFill />  </button>
            <span>{_quantity}</span>
            <button id="btn-add" className="amount-input__button" onClick={handleClickPlus}> <BsFillCaretRightFill /> </button>
        </div>
    )
}

AmountInput.propTypes = {
    quantity: PropTypes.number,
    onSelectQuantity: PropTypes.func.isRequired
};

export default AmountInput;