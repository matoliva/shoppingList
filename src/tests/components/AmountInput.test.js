import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';

import AmountInput from '../../components/AmountInput';

describe('AmountInput Component', () => {

    // It's inicialized in order to have the intellisense.
    let wrapper = shallow(<AmountInput onSelectQuantity={() => { }} />);
    let useEffect;
    let props;

    const mockUseEffect = () => {
        useEffect.mockImplementationOnce(f => f());
    }

    beforeEach(() => {
        useEffect = jest.spyOn(React, "useEffect");

        props = { onSelectQuantity: jest.fn() };

        mockUseEffect();
        mockUseEffect();

        wrapper = shallow(<AmountInput onSelectQuantity={() => { }} />);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('should display <AmountInput /> correctly', () => {

        expect(wrapper).toMatchSnapshot();
    });

    test('should display quantity prop by parameters', () => {
        const quantityMock = 10;
        const wrapper = shallow(<AmountInput quantity={quantityMock} onSelectQuantity={() => { }} />);

        const spanText = wrapper.find('span').text();
        expect(spanText).toBe('10');
    });

    test('should increment by 1 when the user click the +1 button', () => {

        const btnAddElement = wrapper.find('#btn-add');
        btnAddElement.simulate('click');

        const spanText = wrapper.find('span').text();

        expect(spanText).toBe('1');
    });

    test('should decrement by 1 when the user click the -1 button', () => {
        const quantityValue = 1;
        const wrapper = shallow(<AmountInput quantity={quantityValue} onSelectQuantity={() => { }} />);

        const btnSubtractElement = wrapper.find('#btn-subtract');

        btnSubtractElement.simulate('click');
        const spanText = wrapper.find('span').text();

        expect(spanText).toBe('0');
    });

    test('should not subtract when the quantity is 0 and the user click -1 button', () => {

        const btnSubtractElement = wrapper.find('#btn-subtract');

        btnSubtractElement.simulate('click');
        const spanText = wrapper.find('span').text();

        expect(spanText).toBe('0');
    });

    test('should call onSelectQuantity when the user click +1 button', () => {

        wrapper = shallow(<AmountInput {...props} />);
        const btnAddElement = wrapper.find('#btn-add');
        btnAddElement.simulate('click');

        expect(props.onSelectQuantity).toHaveBeenCalledTimes(1);

    });

    test('should call onSelectQuantity when the user click -1 button', () => {

        wrapper = shallow(<AmountInput {...props} />);
        const btnAddElement = wrapper.find('#btn-subtract');
        btnAddElement.simulate('click');

        expect(props.onSelectQuantity).toHaveBeenCalledTimes(1);

    });

});