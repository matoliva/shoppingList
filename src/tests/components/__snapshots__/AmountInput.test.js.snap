// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountInput Component should display <AmountInput /> correctly 1`] = `
<div
  className="amount-input"
>
  <button
    className="amount-input__button"
    id="btn-subtract"
    onClick={[Function]}
  >
     
    <BsFillCaretLeftFill />
      
  </button>
  <span>
    0
  </span>
  <button
    className="amount-input__button"
    id="btn-add"
    onClick={[Function]}
  >
     
    <BsFillCaretRightFill />
     
  </button>
</div>
`;
